### PR TITLE
Add structural module with simple beam finite element model

### LINKED ISSUE #29: Port across BLUEPRINT beams module to bluemira
I will take the opportunity to clean up a bit, and package as many specific utilities inside the module.

This is more of a first pass, and quite a few assumptions are made about the future structure of bluemira, and names of functions and constants. These are however relatively easy to refactor once the architecture is agreed.

This first pass will be a little more dependent on some geometry functionality that may not live forever! Just trying to get some stuff in for now.

### PR DESCRIPTION
**Pull request by CoronelBuendia**
closes #29 

This is an initial pass, in which I have gotten rid of `Loop` and `Shell` from the old `BLUEPRINT` `beams` module and attempted to use primitive geometry (`BluemiraFace`) and `Coordinates` instead.

I have gotten rid of the link to `sectionproperties` for now, meaning that this implementation does not have access to cross-sectional properties calculated by finite element, and only has analytical properties. This is because `sectionproperties` has recently had a major version update, and I wanted to get in at least a basic working version of this before revisiting FE XS properties at a later date.

NOTE: I have left sectionproperties as a dependency because it is used elswhere.

The major downside is that torsional properties (in particular, `J`) does not have a closed form analytical solution.

### CODE REVIEW

**Comment by je-cook** [comment]
When this is updated section-properties has just released v2 and they have removed any version limits on their dependencies in the process

section-properties <v2 limits our scipy version to < 1.6. But v2 and newer require a matplotlib >3.3.4 which we are limited at the moment from #234

**Comment by CoronelBuendia** [comment]
@je-cook @ivanmaione this is ready now. It's a first pass still, but I'd like to get it in so that I can work on a structural model of the TF coils. I will do a clean up along the way; not super happy with it atm. Merge and repent later? :) (it's not used anywhere at present, but will unblock a couple of BLUEPRINT deprecation activities...)

**Comment by CoronelBuendia** [comment]
> There is quite a lot of cleanup that can be done here. Uf we open an issue that as soon as its used it needs to be cleaned up first I am ok with that understanding current time pressures.
> 
> As long as we are aware it needs to be done and that we do it.

It's not urgent; I'm just waiting on other PRs before proceeding with the EU-DEMO build. I have time to take on some clean-up for this. I've spotted a few things already going through it again.. Just didn't want this PR to linger indefinitely at the bottom of the pack

File: bluemira/structural/element.py [hunk 1]
```diff
@@ -0,0 +1,585 @@
...
+            e_mat, g_mat = material.E, material.G
+            properties["EA"] = e_mat * cross_section.area
+            properties["EIyy"] = e_mat * cross_section.i_yy
+            properties["EIzz"] = e_mat * cross_section.i_zz
+            properties["GJ"] = g_mat * cross_section.j
+            properties["ry"] = cross_section.ry
+            properties["rz"] = cross_section.rz
+            properties["rho"] = material.rho
+            properties["A"] = cross_section.area
+            self.material = material
```
Comment (je-cook) [inline bluemira/structural/element.py:273]: I think my indentation is off
```suggestion
        # A composite cross-section was hopefully specified

        properties["nu"] = cross_section.nu
        properties["ry"] = cross_section.ry
        properties["rz"] = cross_section.rz
        properties["A"] = cross_section.area

        if material is None:
            properties["EA"] = cross_section.ea
            properties["EIyy"] = cross_section.ei_yy
            properties["EIzz"] = cross_section.ei_zz
            properties["GJ"] = cross_section.gj
            properties["rho"] = cross_section.rho  # area-weighted density

            # Override material=None with a list of materials from the CS
            self.material = cross_section.material
        else:
            # Single material weight cross-section properties
            e_mat, g_mat = material.E, material.G
            properties["EA"] = e_mat * cross_section.area
            properties["EIyy"] = e_mat * cross_section.i_yy
            properties["EIzz"] = e_mat * cross_section.i_zz
            properties["GJ"] = g_mat * cross_section.j
            properties["rho"] = material.rho
            self.material = material
```

File: bluemira/structural/loads.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    elif load_type == "My":
+        reactions[2] = 6 * load * x * a / length**3
+        reactions[4] = load * a * (2 * x - a) / length**2
+        reactions[8] = -load * x * a / length**3
+        reactions[10] = load * x * (2 * a - x) / length**2
+    elif load_type == "Mz":
+        reactions[1] = -load * x * a / length**3
+        reactions[5] = load * a * (2 * x - a) / length**2
+        reactions[7] = load * x * a / length**3
+        reactions[11] = load * x * (2 * a - x) / length**2
```
Comment (je-cook) [inline bluemira/structural/loads.py:126]: Dont like this (hard to maintain) but probably fine

File: bluemira/structural/model.py [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+        size of the model.
+
+        Parameters
+        ----------
+        deflections: np.array(6*n_nodes)
+            The vector of absolute displacements
+        """
+        xmax, xmin, ymax, ymin, zmax, zmin = self.geometry.bounds()
+        dx, dy, dz = xmax - xmin, ymax - ymin, zmax - zmin
+        length = np.sqrt(dx**2 + dy**2 + dz**2)
```
Comment (je-cook) [inline bluemira/structural/model.py:447]: np.linalg.norm

File: bluemira/structural/stress.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+# You should have received a copy of the GNU Lesser General Public
+# License along with bluemira; if not, see <https://www.gnu.org/licenses/>.
+
+"""
+FE stress interpolations
+"""
+
+import numpy as np
+
+s = np.linspace(0, 1, 10)
```
Comment (je-cook) [inline bluemira/structural/stress.py:28]: this should be a constant, also not used in this file

File: bluemira/structural/transformation.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+            return dcm
+    c11 = np.dot(x_global, x_local)
+    c12 = np.dot(x_global, y_local)
+    c13 = np.dot(x_global, z_local)
+    c21 = np.dot(y_global, x_local)
+    c22 = np.dot(y_global, y_local)
+    c23 = np.dot(y_global, z_local)
+    c31 = np.dot(z_global, x_local)
+    c32 = np.dot(z_global, y_local)
+    c33 = np.dot(z_global, z_local)
```
Comment (je-cook) [inline bluemira/structural/transformation.py:163]: einsum may be quicker here if you get round to it

File: bluemira/structural/transformation.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    -------
+    llambda: np.array((12, 12))
+        The transformation matrix
+    """
+    dcm = _direction_cosine_matrix(dx, dy, dz)
+    llambda = np.zeros((12, 12))
+    llambda[0:3, 0:3] = dcm
+    llambda[3:6, 3:6] = dcm
+    llambda[6:9, 6:9] = dcm
+    llambda[9:12, 9:12] = dcm
```
Comment (je-cook) [inline bluemira/structural/transformation.py:187]: all in one line possible I think with some fancy indexing or use of np.repeat?

**Review by je-cook — COMMENTED** [review]
A first glance through just the main code not the tests

File: bluemira/structural/constants.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+# The proximity tolerance
+#   Used for checking node existence
+D_TOLERANCE = 1e-5
+
+# Small number tolerance
+#   Used for checking if cos / sin of angles is actually zero
+#   Chosen to be slightly larger than:
+#      * np.cos(3 * np.pi / 2) =  -1.8369701987210297e-16
+#      * np.sin(np.pi * 2) = -2.4492935982947064e-16
+NEAR_ZERO = 5e-16
```
[the remark this answers is not in the document]
Comment (je-cook) [reply bluemira/structural/constants.py:51]: actually maybe a factor on EPS

File: bluemira/structural/crosssection.py [hunk 8]
```diff
@@ -0,0 +1,504 @@
...
+    for i in range(len(y) - 1):  # zip is slow in numba
+        y1, y2 = y[i], y[i + 1]
+        z1, z2 = z[i], z[i + 1]
+        d_area = y1 * z2 - y2 * z1
+        q_zz += (y2 + y1) * d_area
+        q_yy += (z2 + z1) * d_area
+        i_zz += (y1**2 + y1 * y2 + y2**2) * d_area
+        i_yy += (z1**2 + z1 * z2 + z2**2) * d_area
+        i_zy += (z1 * y2 + 2 * z1 * y1 + 2 * z2 * y2 + z2 * y1) * d_area
+    return q_zz / 6, q_yy / 6, i_zz / 12, i_yy / 12, i_zy / 24
```
[the remark this answers is not in the document]
Comment (CoronelBuendia) [reply bluemira/structural/crosssection.py:53]: The JIT amounts to the same. Not seen much difference between that and vectorisation in the past for simple loops

File: bluemira/structural/crosssection.py [hunk 9]
```diff
@@ -0,0 +1,504 @@
...
+                y, z = wire.discretize(ndiscr=n_discr, byedges=True).yz
+                q_zz_i, q_yy_i, i_zz_i, i_yy_i, i_zy_i = _calculate_properties(y, z)
+
+                q_zz_o -= q_zz_i
+                q_yy_o -= q_yy_i
+                i_zz_o -= i_zz_i
+                i_yy_o -= i_yy_i
+                i_zy_o -= i_zy_i
+                self.y = np.append(self.y, y)
+                self.z = np.append(self.z, z)
```
[the remark this answers is not in the document]
Comment (CoronelBuendia) [reply bluemira/structural/crosssection.py:422]: Usually not called, and if it is the number of wires is likely never to above 3 or 4

File: bluemira/structural/element.py [hunk 10]
```diff
@@ -0,0 +1,585 @@
...
+            [0, 0, k33, 0, k35, 0, 0, 0, -k33, 0, k35, 0],
+            [0, 0, 0, k44, 0, 0, 0, 0, 0, -k44, 0, 0],
+            [0, 0, k35, 0, k55, 0, 0, 0, -k35, 0, k511, 0],
+            [0, k26, 0, 0, 0, k66, 0, -k26, 0, 0, 0, k612],
+            [-k11, 0, 0, 0, 0, 0, k11, 0, 0, 0, 0, 0],
+            [0, -k22, 0, 0, 0, -k26, 0, k22, 0, 0, 0, -k26],
+            [0, 0, -k33, 0, -k35, 0, 0, 0, k33, 0, -k35, 0],
+            [0, 0, 0, -k44, 0, 0, 0, 0, 0, k44, 0, 0],
+            [0, 0, k35, 0, k511, 0, 0, 0, -k35, 0, k55, 0],
+            [0, k26, 0, 0, 0, k612, 0, -k26, 0, 0, 0, k66],
```
[the remark this answers is not in the document]
Comment (CoronelBuendia) [reply bluemira/structural/element.py:71]: honestly, the diagonal could be done easily but the rest is an indexing nightmare that I don't think would be much more pleasant to look at!

[the remark this answers is not in the document]
Comment (je-cook) [reply bluemira/structural/element.py:71]: fair enough, I was more hopeful than anything

File: bluemira/structural/constants.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+    "Mx": GLOBAL_COORDS[0],
+    "My": GLOBAL_COORDS[1],
+    "Mz": GLOBAL_COORDS[2],
+}
+
+# Mapping of load integer to direction vectors (for plotting purposes)
+LOAD_INT_VECTORS = {k: v for k, v in enumerate(LOAD_STR_VECTORS.values())}
+
+# The length of load vectors relative to space vectors (for plotting purposes)
+NEWTON_TO_M = 5e-6
```
[the remark this answers is not in the document]
Comment (je-cook) [reply bluemira/structural/constants.py:80]: looking back at this maybe that doesnt make sense

File: bluemira/structural/element.py [hunk 12]
```diff
@@ -0,0 +1,585 @@
...
+        self.id_number = id_number
+
+        # Public properties
+        self.loads = []
+
+        # Calculated properties
+        self.shapes = None
+        self.stresses = None
+        self.max_stress = None
+        self.safety_factor = None
```
[the remark this answers is not in the document]
Comment (CoronelBuendia) [reply bluemira/structural/element.py:222]: These are used for cheap caching of properties; sadly can't remove them without a fair bit of effort

File: bluemira/structural/element.py (hunk 1 above)
Comment (CoronelBuendia) [reply bluemira/structural/element.py:273]: doesn't work like this; it's effectively to enable composite properties to be used (which is not the normal case)

File: bluemira/structural/model.py (hunk 3 above)
Comment (CoronelBuendia) [reply bluemira/structural/model.py:447]: image
:stuck_out_tongue_winking_eye: 
(for floats!)

File: bluemira/structural/transformation.py (hunk 5 above)
Comment (CoronelBuendia) [reply bluemira/structural/transformation.py:163]: not used atm. never quite got round to working out why I couldnt get this one to work

File: bluemira/structural/model.py (hunk 3 above)
Comment (je-cook) [reply bluemira/structural/model.py:447]: fair enough!

File: bluemira/structural/transformation.py (hunk 6 above)
Comment (CoronelBuendia) [reply bluemira/structural/transformation.py:187]:
```python
for i in range(4):
    l2[i*3:(i+1)*3, i*3:(i+1)*3] = dcm
```
:S

Comment (je-cook) [reply bluemira/structural/transformation.py:187]: thats worse :smile:

**Review by je-cook — APPROVED** [review]

File: bluemira/structural/transformation.py (hunk 6 above)
Comment (hsaunders1904) [reply bluemira/structural/transformation.py:187]: For future reference, `scipy.linalg` has [`block_diag`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.block_diag.html):

```python
llambda = block_diag(*[dcm]*4)
```

Comment (je-cook) [reply bluemira/structural/transformation.py:187]: Thats the sort of thing I was trying to find!

**Pull request merged**